File: examples/vlm_vision_example.c
/*
 * VLM Vision Plugin Example 1 -- Test source
 * Pipeline: videotestsrc ! videoconvert ! vlmvision ! videoconvert ! ximagesink
 *
 * Build: make all  (from examples/ directory)
 *
 * Run with environment variables:
 *   VLM_API_KEY=sk-... VLM_BASE_URL=https://api.openai.com \
 *     VLM_MODEL=gpt-4o ./bin/vlm_vision_example
 *
 * If no environment variables are set, falls back to a mock server at
 * http://127.0.0.1:8765 with a dummy key.
 */

#include <gst/gst.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <signal.h>

/* ── App data ─────────────────────────────────────────────────── */
typedef struct _AppData {
  GstElement *pipeline;
  GMainLoop *loop;
} AppData;

/* File-scoped copy so the signal handler can reach it. */
static AppData global_data = {NULL, NULL};

/* ── Helpers ──────────────────────────────────────────────────── */

/* ── SIGINT handler ───────────────────────────────────────────── */
static void
on_sigint(int sig)
{
  if (global_data.pipeline) {
    g_print("\n[SIGINT] Sending EOS for graceful shutdown...\n");
    gst_element_send_event(global_data.pipeline, gst_event_new_eos());
  }
}

/* ── "description-received" GObject signal callback ───────────── */
static void
on_description_received(GstElement *element, gchar *description,
                        GstBuffer *buffer, gpointer user_data)
{
  GstClockTime pts = GST_BUFFER_PTS(buffer);
  printf("=================================\n");
  printf("[signal] Frame PTS: %" GST_TIME_FORMAT "\n", GST_TIME_ARGS(pts));
  printf("[signal] Description: %s\n", description);
  printf("=================================\n\n");
  fflush(stdout);
}

/* ── "analysis-error" GObject signal callback ──────────────────── */
static void
on_analysis_error(GstElement *element, gchar *message, gint http_status,
                  gpointer user_data)
{
  printf("*** [signal] ANALYSIS ERROR (HTTP %d): %s ***\n\n",
         http_status, message);
  fflush(stdout);
}

/* ── Bus watch ────────────────────────────────────────────────── */
static gboolean
bus_callback(GstBus *bus, GstMessage *message, gpointer user_data)
{
  AppData *data = (AppData *)user_data;

  switch (GST_MESSAGE_TYPE(message)) {
    case GST_MESSAGE_ERROR: {
      GError *err = NULL;
      gchar *debug = NULL;
      gst_message_parse_error(message, &err, &debug);
      g_printerr("ERROR from %s: %s\n",
                 GST_OBJECT_NAME(message->src), err->message);
      if (debug)
        g_printerr("  Debug info: %s\n", debug);
      g_error_free(err);
      g_free(debug);
      g_main_loop_quit(data->loop);
      break;
    }
    case GST_MESSAGE_EOS:
      g_print("[bus] End of stream.\n");
      g_main_loop_quit(data->loop);
      break;
    case GST_MESSAGE_ELEMENT: {
      const GstStructure *s = gst_message_get_structure(message);
      if (s && gst_structure_has_name(s, "vlmvision-error")) {
        const gchar *msg = gst_structure_get_string(s, "message");
        gint http_status = 0;
        gst_structure_get_int(s, "http-status", &http_status);
        printf("*** [bus] VLM VISION ERROR (HTTP %d): %s ***\n",
               http_status, msg ? msg : "(unknown)");
        fflush(stdout);
      } else if (s && gst_structure_has_name(s, "vlmvision-result")) {
        const gchar *desc    = gst_structure_get_string(s, "description");
        const gchar *model   = gst_structure_get_string(s, "model");
        const gchar *json    = gst_structure_get_string(s, "raw-json");
        const gchar *err_str = gst_structure_get_string(s, "error");
        guint64 pts = 0;
        gst_structure_get_uint64(s, "pts", &pts);

        printf("--- [bus message: vlmvision-result] ---\n");
        printf("  PTS:        %" GST_TIME_FORMAT "\n", GST_TIME_ARGS(pts));
        printf("  Model:      %s\n", model ? model : "(unknown)");
        printf("  Description: %s\n", desc ? desc : "(null)");
        if (err_str && err_str[0])
          printf("  Error:      %s\n", err_str);
        if (json && json[0])
          printf("  Raw JSON:   %zu bytes\n", strlen(json));
        printf("--------------------------------------\n");
        fflush(stdout);
      }
      break;
    }
    case GST_MESSAGE_STATE_CHANGED:
      if (GST_MESSAGE_SRC(message) == GST_OBJECT(data->pipeline)) {
        GstState old, new_, pending;
        gst_message_parse_state_changed(message, &old, &new_, &pending);
        g_print("[bus] Pipeline: %s -> %s\n",
                gst_element_state_get_name(old),
                gst_element_state_get_name(new_));
      }
      break;
    default:
      break;
  }
  return TRUE;
}

/* ── main ─────────────────────────────────────────────────────── */
int main(int argc, char *argv[])
{
  GstElement *pipeline, *source, *conv1, *vlm, *conv2, *sink;
  GstBus *bus;
  AppData data;

  /* ================================================================
   *  1.  Read environment / set defaults
   * ==============================================================*/
  const gchar *api_key  = getenv("VLM_API_KEY");
  const gchar *base_url = getenv("VLM_BASE_URL");
  const gchar *model    = getenv("VLM_MODEL");
  const gchar *profile  = getenv("VLM_PROFILE");

  if (!profile) profile = "openai";

  /*
   * If neither an API key nor a base URL was provided, point at a
   * local mock server so the pipeline can at least start up and
   * demonstrate the mechanics without real credentials.
   */
  if ((!api_key || api_key[0] == '\0') && (!base_url || base_url[0] == '\0')) {
    g_print("WARNING: No API key or base URL set.\n");
    g_print("         Falling back to mock server http://127.0.0.1:8765\n");
    g_print("         (start one with: python3 -m http.server 8765)\n");
    api_key  = "dummy-mock-key";
    base_url = "http://127.0.0.1:8765";
    if (!model) model = "mock-model";
  }

  if (!base_url || base_url[0] == '\0')
    base_url = "https://api.openai.com";

  /* ================================================================
   *  2.  Print configuration summary
   * ==============================================================*/
  g_print("=== VLM Vision Configuration ===\n");
  g_print("  Base URL:       %s\n", base_url);
  g_print("  Model:          %s\n", model);
  g_print("  Profile:        %s\n", profile);
  g_print("  System prompt:  You are a helpful video analysis assistant.\n");
  g_print("  User prompt:    Describe what you see in this video frame in detail.\n");
  g_print("  Output mode:    7 (meta | signal | bus)\n");
  g_print("  Timeout:        10 s\n");
  g_print("  Max inflight:   1\n");
  g_print("  Queue policy:   drop\n");
  g_print("=================================\n");

  /* ================================================================
   *  3.  Initialise GStreamer
   * ==============================================================*/
  gst_init(&argc, &argv);

  /* ================================================================
   *  4.  Create elements
   * ==============================================================*/
  source = gst_element_factory_make("videotestsrc",  "source");
  conv1  = gst_element_factory_make("videoconvert",  "conv1");
  vlm    = gst_element_factory_make("vlmvision",     "vlm-vision");
  conv2  = gst_element_factory_make("videoconvert",  "conv2");
  sink   = gst_element_factory_make("ximagesink", "sink");

  if (!vlm) {
    g_printerr("ERROR: 'vlmvision' element not found.\n");
    g_printerr("       Is GstVlmVision installed? Check GST_PLUGIN_PATH.\n");
    return 1;
  }
  if (!source || !conv1 || !conv2 || !sink) {
    g_printerr("ERROR: Failed to create a basic GStreamer element.\n");
    g_printerr("       Install gst-plugins-base and gst-plugins-good.\n");
    if (vlm)    gst_object_unref(vlm);
    if (source) gst_object_unref(source);
    if (conv1)  gst_object_unref(conv1);
    if (conv2)  gst_object_unref(conv2);
    if (sink)   gst_object_unref(sink);
    return 1;
  }

  /* ================================================================
   *  5.  Configure elements
   * ==============================================================*/
  /* videotestsrc: SMPTE colour bars */
  g_object_set(source, "pattern", 0, NULL);

  g_object_set(vlm,
      "api-key",        api_key,
      "base-url",       base_url,
      "model",          model,
      "profile",        profile,
      "system-prompt",  "You are a helpful video analysis assistant.",
      "user-prompt",    "Describe what you see in this video frame in detail.",
      "output-mode",    7,         /* meta=1 | signal=2 | bus=4 */
      "timeout",        10,
      "max-inflight",   1,
      "queue-policy",   0,         /* 0=drop */
      "error-policy",   1,         /* 1=bus */
      NULL);

  /* Connect signals. */
  g_signal_connect(vlm, "description-received",
                   G_CALLBACK(on_description_received), NULL);
  g_signal_connect(vlm, "analysis-error",
                   G_CALLBACK(on_analysis_error), NULL);

  /* ================================================================
   *  6.  Assemble pipeline
   * ==============================================================*/
  pipeline = gst_pipeline_new("vlm-vision-pipeline");
  data.pipeline = pipeline;
  data.loop     = NULL;

  gst_bin_add_many(GST_BIN(pipeline), source, conv1, vlm, conv2, sink, NULL);

  if (!gst_element_link_many(source, conv1, vlm, conv2, sink, NULL)) {
    g_printerr("ERROR: Failed to link elements.\n");
    gst_object_unref(pipeline);
    return 1;
  }

  /* ================================================================
   *  7.  Bus watch
   * ==============================================================*/
  bus = gst_pipeline_get_bus(GST_PIPELINE(pipeline));
  gst_bus_add_watch(bus, bus_callback, &data);
  gst_object_unref(bus);

  /* ================================================================
   *  8.  Signal handler for Ctrl+C
   * ==============================================================*/
  global_data.pipeline = pipeline;
  global_data.loop     = NULL;
  signal(SIGINT, on_sigint);

  /* ================================================================
   *  9.  Run
   * ==============================================================*/
  data.loop = g_main_loop_new(NULL, FALSE);
  global_data.loop = data.loop;

  g_print("Starting pipeline. Press Ctrl+C to stop.\n");
  gst_element_set_state(pipeline, GST_STATE_PLAYING);
  g_main_loop_run(data.loop);

  /* ================================================================
   *  10. Cleanup
   * ==============================================================*/
  g_print("Cleaning up...\n");
  signal(SIGINT, SIG_DFL);
  gst_element_set_state(pipeline, GST_STATE_NULL);
  gst_object_unref(pipeline);
  g_main_loop_unref(data.loop);

  return 0;
}

File: examples/vlm_vision_example.py
#!/usr/bin/env python3
"""
VLM Vision Plugin -- Python example
Pipeline: videotestsrc ! videoconvert ! vlmvision ! videoconvert ! ximagesink

Demonstrates all major vlmvision plugin capabilities:
  - Environment-variable configuration (API key, base URL, model, profile)
  - Property configuration (profile, prompts, interval, output-mode, timeout, max-inflight)
  - Signal-based result delivery
  - Bus-message-based result delivery
  - Startup configuration summary
  - Graceful error handling and shutdown

Requires: libgstvlmvision.so installed in GStreamer plugin path.
"""

import sys
import os
import gi

gi.require_version("Gst", "1.0")
from gi.repository import Gst, GLib


class VlmVisionApp:
    """Demonstrates the vlmvision GStreamer element with signal and bus output."""

    def __init__(self):
        Gst.init(None)

        # ------------------------------------------------------------------
        # 1. Environment-variable configuration
        # ------------------------------------------------------------------
        api_key = os.environ.get("VLM_API_KEY") or os.environ.get("OPENAI_API_KEY")
        self.base_url = os.environ.get("VLM_BASE_URL")
        self.model = os.environ.get("VLM_MODEL")
        self.profile = os.environ.get("VLM_PROFILE") or "openai"

        if not api_key:
            api_key = ""
        if not self.base_url:
            self.base_url = "https://api.openai.com"
        if not self.model:
            print("WARNING: VLM_MODEL not set. API calls will fail without a valid model name.")

        self.api_key = api_key

        # ------------------------------------------------------------------
        # 2. Prompt & analysis settings
        # ------------------------------------------------------------------
        self.system_prompt = "You are a helpful vision assistant."
        self.user_prompt = "Describe what you see in this image."
        self.analysis_interval = 5.0  # seconds
        self.frames_per_request = 1
        self.temperature = 0.7
        self.max_output_tokens = 200
        self.top_p = 0.9

        # bitmask: 1=meta  2=signal  4=bus  8=raw-json
        # 7 = meta + signal + bus
        self.output_mode = 7
        self.timeout = 30
        self.max_inflight = 2
        self.error_policy = 1   # bus

        # ------------------------------------------------------------------
        # 3. Build pipeline
        # ------------------------------------------------------------------
        self.pipeline = Gst.Pipeline.new("vlm-vision-pipeline")

        source = Gst.ElementFactory.make("videotestsrc", "source")
        conv1 = Gst.ElementFactory.make("videoconvert", "converter1")
        vlm = Gst.ElementFactory.make("vlmvision", "vlm-vision")
        conv2 = Gst.ElementFactory.make("videoconvert", "converter2")
        sink = Gst.ElementFactory.make("ximagesink", "sink")

        for name, elem in [
            ("videotestsrc", source),
            ("videoconvert", conv1),
            ("vlmvision", vlm),
            ("videoconvert", conv2),
            ("ximagesink", sink),
        ]:
            if not elem:
                print(f"ERROR: Could not create element '{name}'.")
                if name == "vlmvision":
                    print(
                        "       Ensure the plugin is installed and "
                        "GST_PLUGIN_PATH is set."
                    )
                sys.exit(1)

        self.source = source
        self.vlm = vlm

        # -- Configure source --
        source.set_property("pattern", 0)  # smpte test pattern

        # -- Configure vlmvision properties --
        vlm.set_property("api-key", self.api_key)
        vlm.set_property("base-url", self.base_url)
        vlm.set_property("model", self.model)
        vlm.set_property("profile", self.profile)
        vlm.set_property("system-prompt", self.system_prompt)
        vlm.set_property("user-prompt", self.user_prompt)
        vlm.set_property("analysis-interval", self.analysis_interval)
        vlm.set_property("frames-per-request", self.frames_per_request)
        vlm.set_property("temperature", self.temperature)
        vlm.set_property("max-output-tokens", self.max_output_tokens)
        vlm.set_property("top-p", self.top_p)
        vlm.set_property("output-mode", self.output_mode)
        vlm.set_property("timeout", self.timeout)
        vlm.set_property("max-inflight", self.max_inflight)
        vlm.set_property("error-policy", self.error_policy)

        # ------------------------------------------------------------------
        # 4. Connect signals
        # ------------------------------------------------------------------
        vlm.connect("description-received", self.on_description_received)
        vlm.connect("analysis-error", self.on_analysis_error)

        # ------------------------------------------------------------------
        # 5. Link elements
        #     videotestsrc ! videoconvert ! vlmvision ! videoconvert ! ximagesink
        # ------------------------------------------------------------------
        for elem in [source, conv1, vlm, conv2, sink]:
            self.pipeline.add(elem)

        if not (
            source.link(conv1)
            and conv1.link(vlm)
            and vlm.link(conv2)
            and conv2.link(sink)
        ):
            print("ERROR: Failed to link pipeline elements.")
            sys.exit(1)

        # ------------------------------------------------------------------
        # 6. Bus (message bus for EOS, errors, and vlmvision-result messages)
        # ------------------------------------------------------------------
        bus = self.pipeline.get_bus()
        bus.add_signal_watch()
        bus.connect("message", self.on_bus_message)

        self.loop = GLib.MainLoop()

    # ------------------------------------------------------------------
    # Signal callback: description-received
    # ------------------------------------------------------------------
    def on_analysis_error(self, element, message, http_status):
        print(f"\n*** [SIGNAL] ANALYSIS ERROR (HTTP {http_status}): {message} ***\n", flush=True)

    def on_description_received(self, element, description, buffer):
        """Called when vlmvision emits the 'description-received' signal."""
        pts = buffer.pts
        if pts != Gst.CLOCK_TIME_NONE:
            pts_str = f"{pts // Gst.SECOND}.{pts % Gst.SECOND:09d}"
        else:
            pts_str = "?"
        print("=" * 50)
        print(f"  [SIGNAL] Frame PTS : {pts_str}s")
        print(f"  [SIGNAL] Description: {description}")
        print("=" * 50, flush=True)

    # ------------------------------------------------------------------
    # Bus message callback
    # ------------------------------------------------------------------
    def on_bus_message(self, bus, message):
        """Process messages from the GStreamer bus."""
        t = message.type

        if t == Gst.MessageType.ERROR:
            err, debug = message.parse_error()
            print(f"\n[ERROR] from {message.src.get_name()}: {err.message}")
            if debug:
                print(f"  Debug: {debug}")
            self.loop.quit()

        elif t == Gst.MessageType.EOS:
            print("\n[EOS] End of stream reached.")
            self.loop.quit()

        elif t == Gst.MessageType.ELEMENT:
            self._handle_element_message(message)

    def _handle_element_message(self, message):
        """Process element-specific bus messages (e.g. vlmvision-result)."""
        structure = message.get_structure()
        if structure is None:
            return

        name = structure.get_name()
        if name == "vlmvision-error":
            msg = structure.get_string("message") or ""
            http_status = 0
            try:
                ok, http_status = structure.get_int("http-status")
                if not ok:
                    http_status = 0
            except Exception:
                pass
            print(f"*** [BUS] VLM VISION ERROR (HTTP {http_status}): {msg} ***", flush=True)
        elif name == "vlmvision-result":
            # Parse fields: description (string), pts (uint64), model (string),
            #               raw-json (string), error (string)
            description = structure.get_string("description")
            model = structure.get_string("model") or ""
            raw_json = structure.get_string("raw-json") or ""
            error = structure.get_string("error") or ""

            # get_uint64 returns (success: bool, value: int) in GI bindings
            pts_val = None
            try:
                ok, pts_val = structure.get_uint64("pts")
                if not ok:
                    pts_val = None
            except Exception:
                # Fallback for binding variants that return a single value
                pass

            print("--- [BUS] vlmvision-result ---")
            if pts_val is not None:
                pts_str = f"{pts_val // Gst.SECOND}.{pts_val % Gst.SECOND:09d}"
                print(f"  PTS        : {pts_str}s")
            if model:
                print(f"  Model      : {model}")
            print(f"  Description: {description}")
            if error:
                print(f"  Error      : {error}")
            if raw_json:
                print(f"  Raw JSON   : ({len(raw_json)} chars)")
            print("--- End bus message ---\n", flush=True)

    # ------------------------------------------------------------------
    # Startup summary
    # ------------------------------------------------------------------
    def _print_config(self):
        """Print a summary of the active configuration."""
        print("=" * 50)
        print(" VLM Vision Pipeline Configuration")
        print("=" * 50)
        print(f"  Base URL         : {self.base_url}")
        print(f"  Model            : {self.model}")
        print(f"  Profile          : {self.profile}")
        print(f"  System prompt    : {self.system_prompt!r}")
        print(f"  User prompt      : {self.user_prompt!r}")
        print(f"  Analysis interval: {self.analysis_interval} s")
        print(f"  Frames/request   : {self.frames_per_request}")
        print(f"  Temperature      : {self.temperature}")
        print(f"  Max output tokens: {self.max_output_tokens}")
        print(f"  Top-p            : {self.top_p}")
        print(f"  Output mode      : {self.output_mode} "
              f"(meta={'1' if self.output_mode & 1 else '0'} "
              f"signal={'1' if self.output_mode & 2 else '0'} "
              f"bus={'1' if self.output_mode & 4 else '0'})")
        print(f"  Timeout          : {self.timeout} s")
        print(f"  Max inflight     : {self.max_inflight}")
        print("-" * 50)
        print("  Pipeline:")
        print("    videotestsrc ! videoconvert ! vlmvision ! videoconvert")
        print("    ! ximagesink")
        print("=" * 50, flush=True)

    # ------------------------------------------------------------------
    # Run
    # ------------------------------------------------------------------
    def run(self):
        """Start the pipeline and run the GLib main loop."""
        self._print_config()
        print("Press Ctrl+C to stop.\n", flush=True)

        ret = self.pipeline.set_state(Gst.State.PLAYING)
        if ret == Gst.StateChangeReturn.FAILURE:
            print("ERROR: Unable to set pipeline to PLAYING state.")
            sys.exit(1)

        try:
            self.loop.run()
        except KeyboardInterrupt:
            print("\n[Interrupt] Received Ctrl+C, shutting down...")
        finally:
            self.pipeline.set_state(Gst.State.NULL)
            print("[Done] Pipeline stopped.")


if __name__ == "__main__":
    VlmVisionApp().run()
